Cel filters -- Review

Harwayne opened the pull request:
DO NOT SUBMIT

Just for review, not meant to merge.

grantr committed:
Add CELExpression field to Trigger.Filter

grantr committed:
Add CEL filtering to Trigger receiver

Triggers with the Filter.CELExpression field populated use CEL
to filter events. The CEL expression has access to a `ce` object with the
following fields:
- `specversion`
- `type`
- `source`

This list of fields is defined by the dev.knative.CloudEventFilterMeta
proto in pkg/broker/dev_knative.

The following CEL expression selects events with type `com.example`:

`ce.type == "com.example"`

grantr committed:
Merge branch 'master' into cel-filters

grantr committed:
Update Trigger CEL fields

Changes the CEL filter to be expressed this way:

filter:
  cel:
    expression: foo.bar = "baz"

grantr committed:
Switch to dynamic types for CEL filters

Using a protobuf Struct eliminates the need to generate pb structs
for CloudEvent fields, and lays the groundwork for adding dynamic fields
to the expression environment.

grantr committed:
Add support for JSON event data

If the event has datacontenttype "application/json", the CEL environment
will have the parsed json data in the data field.

grantr committed:
Add ParseExtensions and ParseData flags

These fields must be true to enable dynamic parsing of event
extensions and data respectively in the CEL environment. If parsing
fails, the standard CE fields will still be available for filtering.

grantr committed:
Log parse and eval errors

Also refactor event context parsing to be more efficient and readable.

grantr committed:
Refactor trigger receiver test to use builders

Triggers and Event fixtures are now generated by builders in an attempt
to make the content of the fixture clearer and more flexible.

grantr committed:
Refactor CEL variables and add tests

The CloudEvents SDK seems unable to return a single object with both the
standard fields and the extension fields. Until that's possible, the
parsed extensions will be prefixed with `ext.` in the CEL expression.

The parsed JSON data will be prefixed with `data.` in the CEL
expression.

The standard fields are currently un-prefixed, but that will probably
need to change because CEL reserves `type`. Currently the CloudEvents
type is called `typ` as a temporary workaround.

Tests now verify that basic expressions on type, source, parsed
extensions, and parsed data filter correctly.

grantr committed:
Prefix CloudEvent context keys

`type` is problematic in CEL because it's a reserved word, and in
general it's probably a bad idea to risk collisions with top-level
reserved words.

This change nests all CE keys under `ce.` to avoid the `type` collision
and all future collisions. A proto is defined for these keys to avoid
issues with dynamic typing and make the definition more shareable.

Eventually the ce and ext prefixes could be combined into a single
prefix for future proofing of extensions and simpler user experience.

grantr committed:
Merge branch 'master' into cel-filters

Harwayne commented on pkg/apis/eventing/v1alpha1/trigger_types.go:
```diff
@@ -65,7 +65,16 @@ type TriggerSpec struct {
 }
 
 type TriggerFilter struct {
+	// SourceAndType filters events based on exact matches on the type and source
```
"type and source attributes" -> "CloudEvents type and source attributes"

Harwayne commented on pkg/apis/eventing/v1alpha1/trigger_types.go:
```diff
@@ -65,7 +65,16 @@ type TriggerSpec struct {
 }
 
 type TriggerFilter struct {
```
Specify that only one of the fields should be set.

Harwayne commented on pkg/apis/eventing/v1alpha1/trigger_types.go:
```diff
@@ -76,6 +85,26 @@ type TriggerFilterSourceAndType struct {
+	// cannot be parsed they will be ignored. Defaults to false.
+	//
+	// +optional
+	ParseExtensions bool `json:"parseExtensions"`
```
omitempty

grantr replied:
I didn't add this for the boolean fields because I thought there might be subtleties with omitted values. After some more research, I think adding `omitempty` is fine as long as the default value for omitted `ParseData` and `ParseExtensions` fields is always `false`.

Harwayne commented on pkg/apis/eventing/v1alpha1/trigger_types.go:
```diff
@@ -76,6 +85,26 @@ type TriggerFilterSourceAndType struct {
+	// JSON data. Defaults to false.
+	//
+	// +optional
+	ParseData bool `json:"parseData"`
```
omitempty

Harwayne commented on pkg/apis/eventing/v1alpha1/trigger_types.go:
```diff
@@ -76,6 +85,26 @@ type TriggerFilterSourceAndType struct {
+	// +optional
+	ParseExtensions bool `json:"parseExtensions"`
+	// ParseData enables parsing of the event data and makes the parsed data
+	// available in the CEL environment. Currently this is only available for
```
Do we think that this will work transparently for other data formats? For example, if the data is XML, will this boolean be sufficient?


Or should we make this an enum, with the current values NONE and JSON, defaulting to NONE.

grantr replied:
The data content type and content encoding are transmitted with the event, so the SDK has everything needed at runtime to decide whether the data is parseable (cc @n3wscott to confirm). I believe a boolean is sufficient here to allow parsing of both JSON and XML events, but I'm open to alternate reasoning.

Harwayne commented on pkg/apis/eventing/v1alpha1/trigger_types.go:
```diff
@@ -76,6 +85,26 @@ type TriggerFilterSourceAndType struct {
+type TriggerFilterCEL struct {
+	// Expression is the CEL expression to evaluate. Required.
+	Expression string `json:"expression,omitempty"`
+	// ParseExtensions enables parsing of dynamic extensions attached to the event
```
What is considered a dynamic extension? Is it anything that is not defined in the CloudEvents spec? 

If so, which version of the spec? If they only ever add extensions, then we can put a minimal version. But if they ever remove an attribute from a subsequent spec version, then if this is false, the user suddenly has a non-working filter, as that extension won't exist.

grantr replied:
Yes, anything not defined in the CloudEvents spec is considered a dynamic extension here. I should clarify that in the comment.

That said, practically the list of "official attributes" available in the CEL environment is defined here, not in the CloudEvents spec. If CloudEvents later removes or renames an attribute, we can keep the original available in the CEL environment to avoid breaking Triggers that previously had `parseExtensions` set to `false`.

Harwayne commented on pkg/broker/cel.go:
```diff
@@ -0,0 +1,137 @@
+)
+
+const (
+	// CELVarKeyContext is the CEL variable key used for the CloudEvent event
```
I've been told that it has to be plural, CloudEvents.

Harwayne commented on pkg/broker/cel.go:
```diff
@@ -0,0 +1,137 @@
+	CELVarKeyExtensions = "ext"
+	// CELVarKeyData is the CEL variable key used for the CloudEvent event data.
+	CELVarKeyData = "data"
+	//TODO add a key that contains both the extensions and the baseline context
```
Space after '//' before TODO.

Harwayne commented on pkg/broker/cel.go:
```diff
@@ -0,0 +1,137 @@
+	CELVarKeyExtensions = "ext"
+	// CELVarKeyData is the CEL variable key used for the CloudEvent event data.
+	CELVarKeyData = "data"
+	//TODO add a key that contains both the extensions and the baseline context
```
This seems very important to me. 

If we don't do this, then what happens when I have running filters while updating to a new Knative eventing release, which happens to use the next version of the CE spec?

I think that would cause all my filters to stop working. In fact, to do upgrade safely, I would need to modify every filter, duplicating the existing logic that uses an extension attribute, changing it to use the standard attribute, then OR those two together.

grantr replied:
This is a great point. The immediate example is subject. As written, a `subject` attribute would be available as `ext.subject`, until this code decides to elevate it to an official attribute at which point it becomes `ce.subject`.

I'd call this a blocker to merging. See the similar discussion at https://github.com/knative/eventing/pull/1047#discussion_r274706752.

grantr replied:
This is fixed by 11710b5829813bb9707426bebfe216c2c5134cad. Extensions, when parsed, are available under the `ce` key.

Harwayne commented on pkg/broker/cel.go:
```diff
@@ -0,0 +1,137 @@
+		//TODO should this coerce to V02?
+		extStruct, err := ceParsedExtensionsStruct(event.Context.AsV02().Extensions)
+		if err != nil {
+			r.logger.Error("Failed to parse event context for CEL filtering", zap.String("id", event.Context.AsV02().ID), zap.Error(err))
```
Shouldn't this return the err?

grantr replied:
There are two options when a parse error occurs:
1. Ignore the error and evaluate the expression anyway with the data we parsed correctly.
1. Return the error without evaluating. The filter will reject any event that fails to parse.

The code currently implements 1, assuming that if evaluation is possible, it should be attempted. Thus the error is not returned here, because that would skip the expression evaluation. It would probably be a good idea to save the error and return it along with the expression evaluation result.

Here's an example of a Trigger that probably wants to evaluate anyway even if extensions cannot be parsed:

```yaml
spec:
  filter:
    cel:
      expression: ce.type == "com.example.event"
      parseExtensions: true
```

A failure parsing extensions might not be what the user expects, since the expression could evaluate correctly regardless of whether extensions are successfully parsed.

If we implement option 2, I'd probably rename the field `MustParseExtensions` or something to indicate that a parse failure causes a rejection of the event.

@n3wscott suggested earlier that it may be possible to detect automatically whether the expression requires parsing of extensions or data by inspecting the CEL AST. This seems like the best case for users, but some users may still want to turn parsing behavior on or off independent of the expression.

One possibly terrible option is to make `parseExtensions` a string field, with initial values `yes` and `no` (the default). Future values could be added: `auto` (when AST inspection is possible, this becomes the default), `required` (to indicate behavior on error).

Harwayne commented on pkg/broker/cel.go:
```diff
@@ -0,0 +1,137 @@
+	if ts.Filter.CEL.ParseData {
+		dataStruct, err := ceParsedDataStruct(event)
+		if err != nil {
+			r.logger.Error("Failed to parse event data for CEL filtering", zap.String("id", event.Context.AsV02().ID), zap.Error(err))
```
Shouldn't this return the err?

Harwayne commented on pkg/broker/dev_knative/ce_context.proto:
```diff
@@ -0,0 +1,12 @@
+  string schemaurl = 4;
+  string datamediatype = 5;
+  string datacontenttype = 6;
+}
```
New line.

Harwayne commented on pkg/broker/receiver.go:
```diff
@@ -212,11 +212,32 @@ func (r *Receiver) getTrigger(ctx context.Context, ref provisioners.ChannelRefer
 		return false
 	}
+
+	// TODO what should happen if multiple filter types are specified? OR? AND?
```
I would say that only one should be allowed. We can always relax this in the future if we want.

grantr replied:
Updated the documentation to reflect this and removed the comment.

Harwayne commented on pkg/broker/receiver.go:
```diff
@@ -212,11 +212,32 @@ func (r *Receiver) getTrigger(ctx context.Context, ref provisioners.ChannelRefer
+		return r.filterEventBySourceAndType(ts, event)
+	}
+	// TODO is this supposed to be default true?
+	return true
```
I don't think it is supposed to get here. Maybe add a log error?

Harwayne reviewed (commented):
Webhook code will also have to change. My recommendation is to allow at most once filter type to be specified. If nothing is specified, then either default in sourceAndType or leave it empty and fail open.

Harwayne committed:
Add documentation for Brokers changing channel provisioners. (#996)

* Add documentation for changing channel provisioners.

* lost -> lose

* Links go to knative.dev.

grantr commented:
/cc @grantr

mattmoor-sockpuppet committed:
Format markdown (#1049)

Produced via: `prettier --write --prose-wrap=always $(find -name '*.md' | grep -v vendor | grep -v .github)`

mattmoor-sockpuppet committed:
Format markdown (#1051)

Produced via: `prettier --write --prose-wrap=always $(find -name '*.md' | grep -v vendor | grep -v .github)`

n3wscott committed:
Update CloudEvents SDK to 0.6.0 (#1052)

grantr committed:
Merge branch 'master' into cel-filters

grantr committed:
Be specific when CloudEvents is assumed

When we have documentation and features that assume events will be
formatted according to the CloudEvents spec, say that clearly in the
docs.

grantr committed:
Use omitempty for boolean fields in Trigger

These fields are optional and default false so they can be safely
omitted.

grantr commented on pkg/broker/cel.go:
```diff
@@ -0,0 +1,137 @@
+
+func ceParsedDataStruct(event *cloudevents.Event) (*structpb.Struct, error) {
+	//TODO CloudEvents SDK might have a better way to do this with data codecs
+	if event.Context.GetDataContentType() == "application/json" {
```
Use `GetMediaType` here.

Harwayne replied:
Did you just review your own code? :stuck_out_tongue_closed_eyes:

grantr replied:
I was pair-reviewing with Scott :smile:

Harwayne committed:
Triggers working via path, rather than host (#1045)

* Initial work on Triggers working via path, rather than host (which requires Isito).

* Move path manipulation functions to a single package.

* Fix the unit tests.

* Add a unit test for the Broker Filter service not being found.

* Update the comment.

* Add a test for mapBrokerToTriggers.

* More unit tests around mapBrokerToTriggers.Map.

* Switch provisioner.ChannelReference to types.NamespacedName.

grantr committed:
Use CloudEvents instead of CloudEvent

grantr committed:
Handle errors in GetDataMediaType

GetDataMediaType now returns an error because it tries to parse the
content type. If an error occurs, log it and use the empty string as the
data media type.

grantr committed:
Format TODOs correctly

Commit:
Add custom reporter for ImmutableField validation errors (#1048)

* Add custom reporter for ImmutableField validation errors

* Upgrades go-cmp to latest commit of master
* Replaces cmp with kmp for producing diffs at runtime
* Adds use of custom reporter for kmp
* Fixes build warnings and errors

* Fix build

* Run codegen

* Grab latest kn/pkg

grantr committed:
Expand the baseline attributes available

Baseline attributes now include ID and Subject, which is a new field
being standardized for 0.3.

The SDK now has common methods for retrieving ID and Time attributes among
others. There's also a single top-level cloudevents package that aliases
all the types in other packages.

akashrv committed:
Remove Istio dependency from Eventing (Part - 1) (#1044)

* WIP

* WIP - In-memory working with E2E tests

* WIP - remove istio dependency from in-memroy channel

* UTs pass, E2E tests pass with in-memory as well as kafka

* fixed uts that failed due to last K8s service change

* Removed unnecessary space from a line

* dding istio annotation to test POD. This will ve needed when running E2E
tests against channels other than in-memory

* Bug fix to set clusterIp of K8s service only when it is not of type ExternalName

* Updated code based on PR comments

* Updates based on PR comments

* Updates based on PR comments

* Fixed UTs

* Updated VENDOR_LICENSE

grantr committed:
Clarify multiple filter strategy behavior

If multiple filter strategies are used, only one is used and others are
ignored. Currently the precedence is CEL, SourceAndType.

grantr committed:
Use DataMediaType to check for parseable data

DataMediaType is the content type with content type options like charset
removed.

chaodaiG committed:
update test-infra (#1056)

chizhg committed:
Fix errors in test readme (#1046)

* fix errors in test readme

* use prettier to format

* fix typos

* fix CR issues

* Do not bother to change the index

chizhg committed:
Run e2e tests in parallel and refactoring (#1042)

* formatting the log in setup and teardown

* formatting the log in setup and teardown

* run the e2e test cases in parallel

* fix compile errors

* labeled the wrong namespace

* resolve the race condition

* fix the namespace error

* fix namespace error

* fix namespace error

* use the default testing namespace everywhere

* Now the test cases are able to run in parallel

* labeled the wrong namespace

* pass the namespace through the command line

* remove the namespace parameter in the function of creating Broker

* use namespace in the flag
'

* still use pkgTest.Flags.Namespace

* still need to print out the resource when cleaning up

* creating one namespace for each test case and refactoring

* fix the build failure and remove useless logs

* minor change

* minor change

* Update README

* fix README

* fix the ko delete stuck problem

* mistakenly deleted something...

* fix build error..

* remove redundant logs

* wait for single pod running instead of all

* resolve CR issues

* fix CR issue

* solve CR issues

* boolean flag cannot use =

* log.Fatal when error happening in flag parsing

* make logger a global variable

Harwayne committed:
Add a Deprecated condition on 'in-memory-channel' Channels. (#1062)

n3wscott committed:
Take advantage of the package alias from cloudevents. (#1054)

* Take advantage of the package alias from cloudevents.

* use error.

* Fix e2e test images.

grantr committed:
Combine all context attributes into the ce var

Combining official and extension context attributes into the same struct
makes standardizing extensions easier since filter expressions don't
have to change.

Also switched to using native maps for CEL typing instead of protobuf
structs. Thanks to @TristonianJones for showing the way.

n3wscott committed:
Reconcile using pkg/controller (#1057)

* Move subscription controller to pkg/controller.

* fix finalizers.,

* merge main funcs.

* need to shuffle some stuff.

* Use pkg rec/testing.

* oh man testing testing testing.

* update deps.

* update deps.

* update deps.

* Found a bug in patchMergeKey for type struct, it has to be a key.

* clean up, adding final test.

* add pkg controller to main.

* fix up contrib.

* Fix main.

* ran codegen.

* cleanup.

* add link to issue to remove patch.

* adding back ref as deprecated ref.

* update codegen.

* remove untrue comment.

* use subscription in comment.

* don't use testing alias, move reconciler out of v1alpha1.

* use logconfig.

* update deps.

TristonianJones commented on pkg/broker/cel.go:
```diff
@@ -57,87 +45,64 @@ func (r *Receiver) filterEventByCEL(ts *eventingv1alpha1.TriggerSpec, event *clo
+		"source":      event.Source(),
+		"subject":     event.Subject(),
+		"id":          event.ID(),
 		// TODO Time. Should this be a string or a (cel-native) protobuf timestamp?
```
Using a `protobuf.Timestamp` would be preferred to a string-based time

TristonianJones commented on pkg/broker/cel.go:
```diff
@@ -57,87 +45,64 @@ func (r *Receiver) filterEventByCEL(ts *eventingv1alpha1.TriggerSpec, event *clo
 		return false, err
 	}
 
```
+1 to caching the `cel.Program` by expression string since they parse / check is considerably more expensive than the evaluation, like orders of magnitude more.

grantr replied:
Done.

Harwayne committed:
Broker is not ready until its constituent pieces are ready. (#1064)

* Broker is not ready until its constituent pieces are ready.

* Fix broker_lifecycle_test.go.

* Fix broker_test.go.

* Increase coverage.

chizhg committed:
Add e2e testing support for gcp-pubsub CCP (#1066)

* add e2e testing for gcp pubsub CCP

* change the variable to local

* 2019, not 2018

* used constants as the CCP name

* delete ccps after tests finish

* add new line for all copyright comments under test...

* shell is not happy with space

* solve all issues except running gcppubsub channel locally

* add debug info for a test case that usually fails

* enable test run on local

* Update test/e2e-tests.sh

Co-Authored-By: Fredy-Z <chizhg@google.com>

* Update test/e2e-tests.sh

Co-Authored-By: Fredy-Z <chizhg@google.com>

* Update test/e2e-tests.sh

Co-Authored-By: Fredy-Z <chizhg@google.com>

* Update test/e2e-tests.sh

Co-Authored-By: Fredy-Z <chizhg@google.com>

* Update test/e2e-tests.sh

Co-Authored-By: Fredy-Z <chizhg@google.com>

* Update test/e2e-tests.sh

Co-Authored-By: Fredy-Z <chizhg@google.com>

* Update test/e2e-tests.sh

Co-Authored-By: Fredy-Z <chizhg@google.com>

* Update test/e2e-tests.sh

Co-Authored-By: Fredy-Z <chizhg@google.com>

* Update test/e2e-tests.sh

Co-Authored-By: Fredy-Z <chizhg@google.com>

* fix CR issues

* remove the test for default

* fix codereview issues

* set E2E_PROJECT_ID globally in this file

Harwayne committed:
Trigger is not ready until its constituent pieces are ready. (#1071)

* Broker is not ready until its constituent pieces are ready.

* Fix broker_lifecycle_test.go.

* Fix broker_test.go.

* Increase coverage.

* Propagate readiness from the Trigger's constituent pieces to the Trigger's readiness.

* Fix unit tests.

* Fix unit tests.

* Remove extra newline.

akashrv committed:
Remove Istio dependency from Eventing (Part - 2) (#1058)

* WIP

* WIP - In-memory working with E2E tests

* WIP - remove istio dependency from in-memroy channel

* UTs pass, E2E tests pass with in-memory as well as kafka

* fixed uts that failed due to last K8s service change

* Removed unnecessary space from a line

* dding istio annotation to test POD. This will ve needed when running E2E
tests against channels other than in-memory

* Bug fix to set clusterIp of K8s service only when it is not of type ExternalName

* WIP kafka channel

* WIP kafka - UTs and E2E pass
More UTs needded

* Updated code based on PR comments

* WIP

* Updates based on PR comments

* Updates based on PR comments

* Fixed UTs

* Updated VENDOR_LICENSE

* WIP. Update fanout sidecar

* Merge from upstream master

* UTs pass, ITs passed. COde ready for PR

* Updates based on PR comments

* Changes based on PR comments

* Added back permission that was removed by mistake

* Remove istio references

* Removed one more reference of istio

* Updates based on PR comments. Ready to merge into master

* Fixed a typo

chizhg committed:
add more logs, try to find the reason for test flakiness (#1075)

n3wscott committed:
Fix imports for sub test. (#1078)

matzew committed:
Fix webhook controller on openshift (#1086)

grantr committed:
Add program caching

CEL programs are LRU cached with the expression string as key. Cache size
is fixed at 100 entries, but we may want to make this configurable.

If an error occurs while generating the program, that error is cached so
the invalid expression isn't parsed for every event.

grantr committed:
Clarify timestamp comment

grantr committed:
Remove protobuf constraint

No longer used for CEL type declarations.

Harwayne committed:
Update Trigger docs to match #1045. (#1055)

chizhg committed:
wait for the default ServiceAccount being created before creating new pods (#1091)

n3wscott committed:
Update Boilerplate to be 2019. (#1088)

Commit:
Issue 734 Move trigger to knative/pkg/controller (#1079)

* Initial changes, compiles sans tests

* merge in pr1071 to trigger.go

* Use the new test format, just one test

* Start the informers...

* after running hack/update-codegen

* address pr comments, use tracker

* run hack/update-codegen

* yeah, enque the right type (trigger), not reconciler... doh

* yeah, enque the right type (trigger), not reconciler... doh. part2

* Add unit tests. update pkg

* remove cruft

* more cruft removed

* address pr comments

n3wscott committed:
Move channel pkg/controller (#1083)

* test working.

* Adding channel  to main.

* clean imports.

* not a todo. removing.

* update deps.

* fix imports from merge.

chizhg committed:
disable gcp pubsub e2e tests until we fix the flaky tests (#1094)

grantr committed:
Remove a few obsolete overrides (#1090)

* Remove cobra and pflag overrides

These overrides are very old, and cobra isn't imported anymore. This
cleanup has already happened in serving.

* Update pflag to 1.0.3

Tracking the serving bump in #2124.

* Remove very old GCP PubSub-related overrides

These were originally added for the PubSub receive adapter in #57, but
it's been a while and we probably don't need them anymore.

matzew committed:
keeping the original deployments as they were, just more on finalizers (#1089)

* keeping the original deployments as they were, just more on finalizers

* use only update

* use only update

n3wscott committed:
Move namespace controller to pkg/controller style. (#1069)

* Start of move namespace controller to pkg/controler style.

* no need for get methods.

* add the namespace controller to the main function.

* working on tests.

* migrate namespace controller to pkg/controller.

* using core informmers.

grantr committed:
Merge branch 'master' into cel-filters

Commit:
add more unit tests for trigger (#1097)

chizhg committed:
update test-infra dep (#1098)

matzew committed:
Kafka dispatcher no need for a statefulset (#972)

* No need for a statefulset

* Feedback from @syedriko

n3wscott committed:
API + Reconciler for CronJobSource. (#1092)

* Move ContainerSource API.

* stage this.

* Move ContainerSource API.

* ported rec.

* stash.

* working cron job source.

* add rbac

* update deps.

* update codegen.

* add deepcopy.

* clean copyright.

* updating to have a source controller.

* fix yaml.

* use sources.

* 2019

* use the real word, not core.

* source -> sources.

* fix yaml.

matzew committed:
k8s forces additionalProperties to true, hence no need to set it ourselves (#1095)

akashrv committed:
Remove Istio dependency from Eventing (Part - 3) - Natss (#1074)

* WIP

* WIP - In-memory working with E2E tests

* WIP - remove istio dependency from in-memroy channel

* UTs pass, E2E tests pass with in-memory as well as kafka

* fixed uts that failed due to last K8s service change

* Removed unnecessary space from a line

* dding istio annotation to test POD. This will ve needed when running E2E
tests against channels other than in-memory

* Bug fix to set clusterIp of K8s service only when it is not of type ExternalName

* WIP kafka channel

* WIP kafka - UTs and E2E pass
More UTs needded

* Updated code based on PR comments

* WIP

* Updates based on PR comments

* Updates based on PR comments

* Fixed UTs

* Updated VENDOR_LICENSE

* WIP. Update fanout sidecar

* Merge from upstream master

* UTs pass, ITs passed. COde ready for PR

* Update natss to not use ISTIO. UTs and E2E tests pass.

* Updates based on PR comments

* REmoved permission to istio virtual service from controller

* Changes based on PR comments

* Added back permission that was removed by mistake

* Remove istio references

* WIP

* Removed one more reference of istio

* Revert kafka.yaml local change

* Revert kafka dispatcher change

* Removing Mutex. No need to use Mutex when using atomic value for hostToChannelMap

* Minor updates based on PR comments

nachocano committed:
Broker controller new infrastructure (#1073)

* first pass to update broker controller

* starting informers properly, listening to own resources

* fix and commenting out UTs

* space

* adding listers

* updating tests

* more tests

* adding deployment functional

* compiling

* more UTs

* more UTs

* more UTs

* more UTs

* more UTs

* updating deps

* updating UTs

* updating UTs

* Updating code gen

* inlining method

* update after comment

* review comments

* serviceInformer and namespaceInformer

* attempt to fix

* solved

* Updates after code review

* Updating gopkg.lock

Harwayne committed:
Add OWNERS for the Channel implementations, based on those who volunteered during the WG call. (#1033)

n3wscott committed:
Move Container source to Eventing, migrate to pkg/controller (#1099)

* Move ContainerSource API.

* stage this.

* Move ContainerSource API.

* ported rec.

* stash.

* working cron job source.

* add rbac

* update deps.

* update codegen.

* add deepcopy.

* clean copyright.

* updating to have a source controller.

* fix yaml.

* use sources.

* 2019

* use the real word, not core.

* source -> sources.

* fix yaml.

* api for container source start.

* rec done first pass. tests next.

* stagging for test writing.:

* first test.

* more tests.

* working tests.

* rbac.

Commit:
make sure NewController doesn't return nil (#1103)

matzew committed:
Consistency is king, use "" for the filter (#1102)

shashwathi committed:
Add data validation on subscription reconciler (#1100)

Fixes https://github.com/knative/eventing/issues/1081

- Update subscription error when reply strategy does not contain address

nachocano committed:
Event Registry (#1105)

* Adding some tests to trigger

* More UTs

* More UTs

* More UTs

* Namespace reconciler automatically creates the Broker Filter's ServiceAccount and RBAC.

Sadly this doesn't work well because we have such an old version of controller-runtime that the Filter ends up trying to watch _all_ Triggers, not just those in its namespace. And it only gets permission for the Triggers in its own namespace.

* Remove no longer needed label.

* Broker and trigger types UTs

* WIP early E2E test

* Changes after code review. Adding trigger defaults and validation tests

* Cleaner trigger validation

* Adding dummy tests for broker validation...
Should be implemented

* Compiling and moving things around

* Updating test

* More updates

* Waiting for potentially multiple contents.
Removing check for corev1.Service ready.
Removing grant's great design. Just making it simpler for now.

* Compiling

* Fixing compilation

* Fixing compilation errors. Adding AnnotateNamespace function.

* Adding ns

* Adding logs. Changing to lowercase any otherwise the pod name is invalid

* Removing namespace when creating trigger subscriber spec.
Adding wait time constant for default broker creation.

* Checking if all triggers are ready

* Updated logs

* Working

* Adding logs... Still not receiving the events.

* More logs

* Adding build constraint

* Removing unnecessary stuff

* Removing ugly structs

* More logs

* Removing quotes

* More logs

* Adding delay

* Listing triggers in receiver when we create it, so that we don't miss
any message because the client couldn't find the existing trigger.
This is a problem in the in-memory-channel as it doesn't do retries.
Maybe the right solution is to add that functionality there.

* Adding delay to sender pod

* Removing withDelay method and just sleep for a while

* Improve log...

* Updates after code review.

* Adding some more logs and trailing dots.

* Switch import order.

* Updating comments.

* Updating comments.

* Replace the bad errgroup usage with the runnableServer.

* Namespace scoped the Broker Filter's client.

* Fix unit tests.

* Fix yaml

* Setting source to source not type.
Updating comment.

* Switch from annotating the namespace to labeling it, to match Istio.

* Adding EventType CRD

* Update to type

* Updates

* updating eventtypes to event-types

* Ingress policy

* Updating environment variables

* yamls to give event-type permissions

* General clean up.

* Adding namespace to ingress and reconciling service accounts in ns.

* More changes

* Service account in namespace

* Updating events registered

* Updating origin to source.

* to newer version

* moving the ingress policy to the broker itself...
should be an object instead of a string.

* making header a constant. Should use the SDK instead.

* Adding dummy change to trigger event

* Using cloudevents sdk but still needs to be updated.
We should change our APIs to receive Events instead of Messages.

* Adding cloudevents-sdk library

* Invalid eventtype

* Initial docs for the Broker.

* Fill some of the usage section.

* Add instructions for installing the Broker without using Namespace annotation.

* Create example_{brokers,triggers}.yaml to document how they can be used.

* Fix MD linter issues.

* Fix MD linter issues.

* Fix MD linter issues.

* Minor clean up.

* adding type as authoritative

* Should actually compare to lowercase

* listing all event types... we should use some sort of selectors.
label selectors won't work on spec.type as they may not conform to the
label format.
field selectors doesn't seem to work for CRDs beyond metadata.name and
namespace.

* Might not need origin after all, but adding it here

* adding origin to yaml as well

* eventtypes instead of event-types to maintain consistency.

* Clean up some spots the merge didn't catch.

* updating example

* Removing source

* cluster-scoped registry to simplify things.

* updating example

* removing source

* updating broker policy to create event types...
updating cloud event SDK version.

* adding a TODO to update to a cluster-level role binding for the ingress.
If we are planning on using cluster-scoped EventTypes

* namespaced EventTypes, i.e., registry per namespace

* adding create permission to ingress

* registry per namespace. Ingress policies working!

* Fix the bad merge by replacing logger.BaseLogger with logger.FormatLogger.

* Add extra columns when using kubectl get.

* MarkBrokerDoesNotExist

* Rename extra columns.

* cosmetic change

* Replace the Trigger reconciler's in-memory map with a simple list, utilizing the fact that Controller Runtime already has the results of the List cached.

* Accept v0.1 and v0.2 cloud events. Adding UTs.
Updating initClient as well, removing unnecessary paging.

* Change to resolve.SubscriberSpec().

* Remove restClient as it wasn't actually used.

* Only reconcile the Namespace if the specific resource we care about changes.

* changing origin to from

* lowercase

* Merge remote-tracking branch 'upstream/master' into registry-namespace

# Conflicts:
#	cmd/broker/ingress/main.go
#	cmd/webhook/main.go
#	config/200-broker-clusterrole.yaml
#	config/500-controller.yaml
#	pkg/apis/eventing/v1alpha1/broker_types.go
#	pkg/apis/eventing/v1alpha1/register.go
#	pkg/reconciler/v1alpha1/broker/broker.go
#	pkg/reconciler/v1alpha1/broker/resources/ingress.go
#	pkg/reconciler/v1alpha1/namespace/namespace.go
#	pkg/reconciler/v1alpha1/trigger/trigger.go

* updating event type to include broker

* additional columns

* just TODOs

* adding some TODOs

* Adding TODOs, updating cloudevents after my change

* Adding observability

* validation

* broker default

* event type controller... checks for broker existence and readiness

* adding extra columns

* message in columns

* Passing broker. Need to refactor this module

* broker immutable to avoid problem with label selector

* listening for broker changes

* removing broker label. Making broker mutable again.

* fixing dns subdomain

* ingress policy as an object.
adding some UTs.
still not convinced.

* removing DeprecatedGeneration

* Adding README

* yaml

* updating readme

* updating trigger

* all fields immutable in eventType.
this makes our life easier when reconciling them from sources.

* adding UTs

* Updating README for source

* moving ToDNSSubdomain to utils

* removing ommitempty

* Using github.com/kelseyhightower/envconfig for env variables.

* changing Triggers to also look for an extension attribute as part of its
source. If there, then we use that as source for exact matching.

* Updating ingress to read From custom extension, if present, and use that
Otherwise it uses source

* Updating comments

* It seems to complaint if I add an invalid URL in the markup.
Changing schema.

* typo

* lowercase from

* using new cloud event API

* addressing Grant's comment.
Debug instead of Info

* Bumping cloud-sdk

* Migrate to correct sdk usage.

* update README with other branch

* updating e2e tests

* Updating deepcopy

* naming changes

* disabling registry test to see if this is causing problems

* re-enabling these guys again

* Missing files

* adding description

* removing autoAdd

* some updates, still broken

* some updates, still broken

* removing broker ingress policies related stuff

* adding eventType controller

* eventType controller new serving structure

* bug

* removing commented code

* adding Gopkg.lock

* rollback source getter

* permissions!!! damn...

* permissions.

* removing formatting of imports in goland

* tracking changes to broker as trigger does.

akashrv committed:
Remove Istio dependency from Eventing (Part - 4) - gcp-pubsub (#1107)

* WIP

* WIP - In-memory working with E2E tests

* WIP - remove istio dependency from in-memroy channel

* UTs pass, E2E tests pass with in-memory as well as kafka

* fixed uts that failed due to last K8s service change

* Removed unnecessary space from a line

* dding istio annotation to test POD. This will ve needed when running E2E
tests against channels other than in-memory

* Bug fix to set clusterIp of K8s service only when it is not of type ExternalName

* WIP kafka channel

* WIP kafka - UTs and E2E pass
More UTs needded

* Updated code based on PR comments

* WIP

* Updates based on PR comments

* Updates based on PR comments

* Fixed UTs

* Updated VENDOR_LICENSE

* WIP. Update fanout sidecar

* Merge from upstream master

* UTs pass, ITs passed. COde ready for PR

* Update natss to not use ISTIO. UTs and E2E tests pass.

* Updates based on PR comments

* REmoved permission to istio virtual service from controller

* WIP

* Changes based on PR comments

* Added back permission that was removed by mistake

* WIP

* Remove istio references

* WIP

* Removed one more reference of istio

* Revert kafka.yaml local change

* WIP

* Revert kafka dispatcher change

* Removing Mutex. No need to use Mutex when using atomic value for hostToChannelMap

* Removed named port from GCP dispatcher K8s service

* WIP

* FInal changes before validating E2E tests

* Updates based on PR comments

syedriko committed:
compile-time assert numControllers == len(controllersArray) (#1112)

* compile-time assert numControllers == len(controllersArray)

* Incorporated review feedback from @grantr

* Typo fix

syedriko committed:
Fix for https://github.com/knative/eventing/issues/1109 (#1118)

[BUG] eventing controller and source-controller fall off main() on shutdown signal

mattmoor-sockpuppet committed:
golang format tools (#1123)

Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -type f -name '*.go' -print))`
  `goimports -w $(find -name '*.go' | grep -v vendor)`

mattmoor-sockpuppet committed:
Fix spelling errors (#1124)

Produced via: `github.com/client9/misspell`

chizhg committed:
Enable e2e testing for NATSS (#1126)

* clean scripts

* update test script

* enable e2e for natss provisioner

* clean up

* fix CR issues

nachocano committed:
Namespace Controller watching for Broker, ServiceAccount, and RoleBinding (#1127)

* Reconciling broker, service account, and role binding in Namespace controller

* updating Gopkg.lock

lionelvillard committed:
move apiserver source to eventing (#1108)

mattmoor-sockpuppet committed:
golang format tools (#1132)

Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -type f -name '*.go' -print))`
  `goimports -w $(find -name '*.go' | grep -v vendor)`

akashrv committed:
Remove Istio dependency from Eventing (Part - 5) - final cleanup (#1119)

* WIP

* WIP - In-memory working with E2E tests

* WIP - remove istio dependency from in-memroy channel

* UTs pass, E2E tests pass with in-memory as well as kafka

* fixed uts that failed due to last K8s service change

* Removed unnecessary space from a line

* dding istio annotation to test POD. This will ve needed when running E2E
tests against channels other than in-memory

* Bug fix to set clusterIp of K8s service only when it is not of type ExternalName

* WIP kafka channel

* WIP kafka - UTs and E2E pass
More UTs needded

* Updated code based on PR comments

* WIP

* Updates based on PR comments

* Updates based on PR comments

* Fixed UTs

* Updated VENDOR_LICENSE

* WIP. Update fanout sidecar

* Merge from upstream master

* UTs pass, ITs passed. COde ready for PR

* Update natss to not use ISTIO. UTs and E2E tests pass.

* Updates based on PR comments

* REmoved permission to istio virtual service from controller

* WIP

* Changes based on PR comments

* Added back permission that was removed by mistake

* WIP

* Remove istio references

* WIP

* Removed one more reference of istio

* Revert kafka.yaml local change

* WIP

* Revert kafka dispatcher change

* Removing Mutex. No need to use Mutex when using atomic value for hostToChannelMap

* Removed named port from GCP dispatcher K8s service

* WIP

* FInal changes before validating E2E tests

* Removed istio sidecars from all dispatchers and from broker.

* Updates based on PR comments

* gcppubsub istio/serviceentry removed

* cronjob source tested with an without sidecar

* Container source

* Removed sidecars from tests

* hack/update-dep.sh

* Fixed build error

* Updates based on PR comments

grantr committed:
Merge branch 'master' into cel-filters

Harwayne committed:
Remove unneeded log keys. (#1135)

* Remove unneeded log keys.

knative.dev/controller is set by pkg/reconciler just before calling our reconciler, so adding a second one in main is redundant and produced entries with:
{
    "knative.dev/controller": "controller",
    "knative.dev/controller": "trigger-controller",
}

The reconcile key is already saved as 'knative.dev/key' by pkg/reconciler, so we don't need to add 'key' as well.

* Remove explict adds of key.

* PR comments.

grantr committed:
Upgrade knative/pkg for new stackdriver exporter (#1139)

The latest knative/pkg upgrades its dependency on the stackdriver
exporter to v0.9.2, which is now compatible with protobuf 1.3.

grantr committed:
Merge branch 'master' into cel-filters
